engine/src/main/java/org/hibernate/search/engine/search/reference/FieldReference.java

```diff
@@ -0,0 +1,38 @@
+package org.hibernate.search.engine.search.reference;
+
+import org.hibernate.search.engine.search.common.ValueConvert;
+import org.hibernate.search.util.common.annotation.Incubating;
+
+/**
+ * @param <T> The expected returned type.
+ */
+@Incubating
+public interface FieldReference<T> {
```

Comment:
Did you look into how the static metamodel in JPA works? If not, I think you might want to; there are a few things we might want to take inspiration from.

One aspect that, I think, we must consider right away is the type holding the field. In JPA you have `SingularAttribute<X, T>`, where `T` is the type of the attribute, and `X` is the _type holding the attribute_ (entity, embeddable).

In our case, implementing things like this raises a few questions:

* For fields within object fields, what is `X`? Does it even make sense?
* In Hibernate Search we often use chains of fields, i.e. `authors.name`. Would it be possible to continue to do this in a way that is not overly verbose? I.e. by default `Book_.authors.name` would have `X = Book`, and there would be a syntax to make the reference relative, e.g. `Book_.authors.relative(a -> a.name)` or `Book_authors.name.relative()` (simpler but quite limited) or ...
* Would it even be possible for us to take advantage of `X`, i.e. to require that one only ever use a field in a context where the parent type/object field is `X`? Would it be possible without breaking APIs (I'm thinking of things such as `org.hibernate.search.engine.search.predicate.dsl.GenericBooleanPredicateClausesStep#should(java.util.function.Function<? super org.hibernate.search.engine.search.predicate.dsl.SearchPredicateFactory,? extends org.hibernate.search.engine.search.predicate.dsl.PredicateFinalStep>)`)?
   Related: `org.hibernate.search.engine.search.predicate.dsl.SearchPredicateFactory#withRoot`.

Reply:
Hey, thanks for taking a look!!! 😃 

yeah, I was looking a bit at the model generated by the ORM. And I thought that it probably can be a good next step to validate some of these assumptions for the Search model. For now, I was thinking about something along the lines:

```java
public abstract class MySearchEntity_ {
	//I'm not sure that we need the `MySearchEntity` here, but maybe we can make use of it in combination with the scope? i.e. we know the most common scope type and do some compile time checks to make sure that things match? 

	public static volatile SearchValueFieldReference<MySearchEntity, Long, String> id;
	public static volatile SearchValueFieldReference<MySearchEntity, LocalDateTime, Instant> dateTime;
	public static volatile SearchValueFieldReference<MySearchEntity, String, String> string;
	public static volatile SearchValueFieldReference<MySearchEntity, Integer, Integer> integer;
	public static volatile SearchValueFieldReference<MySearchEntity, SomeEnum, String> enuM;

	public static volatile MySearchEntityNestedObject_ nested;
}

public abstract class MySearchEntityNestedObject_ extends SearchObjectFieldReference<? not sure what or if needed at all> {
	public static volatile SearchValueFieldReference<MySearchEntityNestedObject, String, String> string;
	// ... 
}
```

So, the classes that are embedded "implement" the object reference interface, and we simply add them to the generated model. That way we'd be able to do something like `MySearchEntity_.nested.string` . We'd need to be careful to address the cycles 🙈 

I'm thinking this way ^ we'd be able to say `f -> f.nested(MySearchEntity_.nested)` as well as `f -> f.match().field(MySearchEntity_.nested.string)`

I would probably prefer if we could make these "paths" to be simply chained through the dots... as path strings are 😃 and don't do method calls if possible.

Reply:
> ```java
> 	//I'm not sure that we need the `MySearchEntity` here, but maybe we can make use of it in combination with the scope? i.e. we know the most common scope type and do some compile time checks to make sure that things match? 
> ```

Yes that's what I was suggesting. However, that would require adding generic type parameters to e.g. `SearchPredicateFactory` and resulting DSL steps, so that you can restrict what type of fields can be used:

>     * Would it even be possible for us to take advantage of `X`, i.e. to require that one only ever use a field in a context where the parent type/object field is `X`? Would it be possible without breaking APIs (I'm thinking of things such as `org.hibernate.search.engine.search.predicate.dsl.GenericBooleanPredicateClausesStep#should(java.util.function.Function<? super org.hibernate.search.engine.search.predicate.dsl.SearchPredicateFactory,? extends org.hibernate.search.engine.search.predicate.dsl.PredicateFinalStep>)`)?
>       Related: `org.hibernate.search.engine.search.predicate.dsl.SearchPredicateFactory#withRoot`.

Alternatively, we could consider an alternative entry point into the DSL when using field references.

E.g. instead of:

```java
f.match().field("foo").matching("sometext")
f.match().field("foo").boost(2.0f).field("bar").matching("sometext")
```

People would do:

```
f.field(MySearchEntity_.foo).matching("sometext")
f.field(MySearchEntity_.foo).boost(2.0f).field(MySearchEntity_.bar).matching("sometext")
```

Pro: more natural syntax, closer to what we do for sorts/projections; shorter syntax (see how there's no `match()` anymore?); no breaking change required.
Con: we now have two relatively different syntaxes in the DSL... And I'd personally feel bad deprecating the old one given the big change we did in 6.0. But maybe we can live with it?



> ```java
> public abstract class MySearchEntityNestedObject_ extends SearchObjectFieldReference<? not sure what or if needed at all> {
> ```

Quite. Given object fields do not necessarily have a Java equivalent, what would we put in the generic type argument?
Related: do we want to use these references for projections, and if so what would be the type of the resulting projection?



> I'm thinking this way ^ we'd be able to say `f -> f.nested(MySearchEntity_.nested)` as well as `f -> f.match().field(MySearchEntity_.nested.string)`

Sure, makes sense.

However, ideally this should *also* allow us to do `f.withRoot(MySearchEntity_.nested).match().field(MySearchEntityNestedObject_.string)`. I'm not saying people *must* do this, just that they should be able to do it.